File: csharp-blazor-webgl/Lib/Math/Vector4.cs
using System.Numerics;

namespace BlazorExperiments.Lib.Math;

public struct Vector4<T> :
    IAdditionOperators<Vector4<T>, Vector4<T>, Vector4<T>>,
    ISubtractionOperators<Vector4<T>, Vector4<T>, Vector4<T>>,
    IMultiplyOperators<Vector4<T>, T, Vector4<T>>,
    IDivisionOperators<Vector4<T>, T, Vector4<T>>,
    IUnaryNegationOperators<Vector4<T>, Vector4<T>>
    where T : INumber<T>
{
    public readonly T X;
    public readonly T Y;
    public readonly T Z;
    public readonly T W;

    public Vector4(T x, T y, T z, T w)
    {
        X = x;
        Y = y;
        Z = z;
        W = w;
    }

    public override string ToString()
    {
        return $"({X}, {Y}, {Z}, {W})";
    }

    public static Vector4<T> operator +(Vector4<T> left, Vector4<T> right)
    {
        return new(
            left.X + right.X,
            left.Y + right.Y,
            left.Z + right.Z,
            left.W + right.W
        );
    }

    public static Vector4<T> operator -(Vector4<T> left, Vector4<T> right)
    {
        return new(
            left.X - right.X,
            left.Y - right.Y,
            left.Z - right.Z,
            left.W - right.W
        );
    }

    public static Vector4<T> operator *(Vector4<T> left, T right)
    {
        return new(
            left.X * right,
            left.Y * right,
            left.Z * right,
            left.W * right
        );
    }

    public static Vector4<T> operator /(Vector4<T> left, T right)
    {
        return new(
            left.X / right,
            left.Y / right,
            left.Z / right,
            left.W / right
        );
    }

    public static Vector4<T> operator -(Vector4<T> value)
    {
        return new(-value.X, -value.Y, -value.Z, -value.W);
    }

    public T MagnitudeSquared => X * X + Y * Y + Z * Z + W * W;

    public T DotProduct(Vector4<T> other)
    {
        return X * other.X + Y * other.Y + Z * other.Z + W * other.W;
    }
}

public static class Vector4Extensions
{
    public static T GetMagnitude<T>(this Vector4<T> v) where T : INumber<T>, IRootFunctions<T>
    {
        return T.Sqrt(v.MagnitudeSquared);
    }

    public static Vector4<T> Normalized<T>(this Vector4<T> v) where T : INumber<T>, IRootFunctions<T>
    {
        return v / v.GetMagnitude();
    }
}

File: swift-c-interop-sdl3-wgpu/Sources/Experiment/math-and-graphics-types/Vector4.swift
struct Vector4<T> {
    let x: T
    let y: T
    let z: T
    let w: T
}

extension Vector4 where T: Mathable {
    static func + (left: Self, right: Self) -> Self {
        Self(
            x: left.x + right.x,
            y: left.y + right.y,
            z: left.z + right.z,
            w: left.w + right.w
        )
    }

    static func - (left: Self, right: Self) -> Self {
        Self(
            x: left.x - right.x,
            y: left.y - right.y,
            z: left.z - right.z,
            w: left.w - right.w
        )
    }

    static func * (left: Self, right: T) -> Self {
        Self(
            x: left.x * right,
            y: left.y * right,
            z: left.z * right,
            w: left.w * right
        )
    }

    static func * (left: T, right: Self) -> Self {
        Self(
            x: left * right.x,
            y: left * right.y,
            z: left * right.z,
            w: left * right.w
        )
    }

    static func / (left: Self, right: T) -> Self {
        Self(
            x: left.x / right,
            y: left.y / right,
            z: left.z / right,
            w: left.w / right
        )
    }

    static prefix func + (unary: Self) -> Self {
        unary
    }

    static prefix func - (unary: Self) -> Self {
        Self(
            x: -unary.x,
            y: -unary.y,
            z: -unary.z,
            w: -unary.w
        )
    }

    static func dot(_ left: Self, _ right: Self) -> T {
        left.x * right.x + left.y * right.y + left.z * right.z + left.w * right.w
    }

    var magnitudeSquared: T { x * x + y * y + z * z + w * w }
}

extension Vector4 where T: Mathable & Sqrt {
    var magnitude: T { magnitudeSquared.sqrt }

    var normalized: Self { self / magnitude }
}
